# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#'
#' \code{get_nlines} returns number of lines. Last line, if empty, is ignored.
#' 
#' @param fn Filename.
#' @export
#' @return \code{get_nlines}: Number of lines, or \code{0} on error.
#' @rdname auxfunc
#' @aliases get_nlines
#' @name get_nlines
#' @include auxil.R
get_nlines <- function(fn) {
    .Call('_Siccuracy_get_nlines', PACKAGE = 'Siccuracy', fn)
}

#' Column bind files with genotype matrices
#'
#' Does row-wise concatenation of multiple genotype files, 
#' ignoring the first column in subsequent files.
#' \strong{NB!} Assumes rows are ordered identically in all files, but will check that
#' value in first column is identical. If not, corresponding line is skipped in all files.
#' 
#' @param fnin Character vector of filenames to concatenate. 
#' @param fnout Filename of resulting file.
#' @param skiplines Integer, number of lines to skip before outputting.
#' @param excludeids Integer vector of first column id's to \emph{exclude} from the output. 
#' @inheritParams convert_phases
#' @return Number of lines written.
#' @rdname cbind_snp_files
#' @backref src/bind_snps.cpp
#' @seealso \code{\link{rbind_snp_files}}
#' @export
cbind_snp_files <- function(fnin, fnout, skiplines = 0L, excludeids = as.integer( c()), idwidth = 4L, precision = -1L) {
    .Call('_Siccuracy_cbind_snp_files', PACKAGE = 'Siccuracy', fnin, fnout, skiplines, excludeids, idwidth, precision)
}

#' Convert phase files to genotype files.
#' 
#' Simply sums every pair of rows together.
#' 
#' A phase file has format similar to SNP files, expect the genotype on each allele are listed on two pairs of rows.
#' Each individual has therefore two rows, one for each allele, see following example:
#'
#' \code{Genotype file:}
#' \tabular{lccccc}{
#'  1003 \tab 0 \tab 1 \tab 1 \tab 2 \tab 0 \cr
#'  1004 \tab 1 \tab 1 \tab 0 \tab 1 \tab 0 \cr
#' }
#'
#' \code{Phase file:}
#' \tabular{lccccc}{
#'  1003 \tab 0 \tab 0 \tab 1 \tab 1 \tab 0 \cr
#'  1003 \tab 0 \tab 1 \tab 0 \tab 1 \tab 0 \cr
#'  1004 \tab 0 \tab 1 \tab 0 \tab 0 \tab 0 \cr
#'  1004 \tab 1 \tab 0 \tab 0 \tab 1 \tab 0 \cr
#' }
#'
#' \strong{Missing values:} Values after summing less than 0 or greater than 2 are assumed as missing and replaced with \code{na}.
#' Value of this range can be changed with argument \code{range}.
#'
#' @param fnin Filename of input file, every two rows are for same animal.
#' @param fnout Filename of output file.
#' @param nlines Integer, maximum number of pairs of lines to convert. 
#'               When \code{-1} (default), no maximum.
#' @param na Value to use for missing values.
#' @param range Integer vector of range of allowable values. 
#'              Values in either input or output that are strictly smaller than 
#'              first element or strictly larger than second element are replaced
#'              with \code{na}.
#' @param idwidth Width of ID column.
#' @param precision Number of decimals to print genotypes; 
#'                  when \code{-1} (default) genotypes are printed with mixed 
#'                  precision. When \code{0}, genotypes are rounded to whole integers.
#' @return Number of rows written.
#' @backref src/convert_phases.cpp
#' @export
convert_phases <- function(fnin, fnout, nlines = -1L, na = 9L, range = as.integer( c(0, 2)), idwidth = 4L, precision = 0L) {
    .Call('_Siccuracy_convert_phases', PACKAGE = 'Siccuracy', fnin, fnout, nlines, na, range, idwidth, precision)
}

#' Convert PLINK recoded A to SNP file.
#' 
#' Facilitates converting a PLINK binary file to simplified SNP file format.
#' Requires using PLINK to recode it to the \code{A} format by using command line \code{plink -bfile <file stem> --recode A}.
#' This function then swiftly strips of first 6 columns (family ID, sample ID, paternal ID, maternal ID, sex, phenotypic record) 
#' and inserts an integer-based ID column. \code{NA}'s outputted from PLINK are replaced with \code{na} argument.
#' 
#' @param fnraw Plink output filename. Most likely \code{plink.raw} if PLINK command line argument \code{--out} is not used.
#' @param fnout Filename of new file.
#' @param newID Integer scalar (default \code{1}) for first integer ID, or vector of new integer IDs.
#' @param na Character. Numeric value to use for missing values, but passed as a character scalar.
#' @inheritParams convert_phases
#' @return Data.frame with columns \code{famID}, \code{sampID}, and \code{newID}.
#' @references 
#' \itemize{
#'  \item PLINK. Purcell and Chang. \url{https://www.cog-genomics.org/plink2}
#'  \item Chang CC, Chow CC, Tellier LCAM, Vattikuti S, Purcell SM, Lee JJ (2015) Second-generation PLINK: rising to the challenge of larger and richer datasets. GigaScience, 4. doi: \href{https://doi.org/10.1186/s13742-015-0047-8}{10.1186/s13742-015-0047-8} \href{http://gigascience.biomedcentral.com/articles/10.1186/s13742-015-0047-8}{link}.
#' }
#' @export
#' @backref src/convert_plinkA.cpp
#' @seealso 
#' \code{\link{convert_plink}} is a direct conversion that does not rely on PLINK.
convert_plinkA <- function(fnraw, fnout, newID = as.integer( c(1)), nlines = -1L, na = "9", idwidth = 4L, precision = 0L) {
    .Call('_Siccuracy_convert_plinkA', PACKAGE = 'Siccuracy', fnraw, fnout, newID, nlines, na, idwidth, precision)
}

